content/authentication/next-steps/bonus-module/spring-security/_index.md

```diff
@@ -0,0 +1,731 @@
+which has a method `authenticate()`. We will create a Username/Password token
+that we send to the `authenticate()` method for username/password validation.
+We will also use the `securityContextRepository` field to save the new
+`SecurityContext` established by the user login. If we catch an
+`AuthenticationException` instance, we will let the user know that the
+username or password were incorrect.
+
+Add the following block to your `processLoginForm` method:
+
+```java
+        try {
+            UsernamePasswordAuthenticationToken token =
+                UsernamePasswordAuthenticationToken.unauthenticated(
+                    loginFormDTO.getUsername(),
+                    loginFormDTO.getPassword()
+                );
+            Authentication authentication =
+                authManager.authenticate(token);
+
+            SecurityContext context = SecurityContextHolder.createEmptyContext();
+            context.setAuthentication(authentication);
+            this.securityContextRepository.saveContext(context, request, response);
+
+            return "redirect:";
+        } catch (AuthenticationException ex) {
+            errors.rejectValue("username", "bad.credentials", "Invalid e-mail or password");
+            return "/login";
+        }
+```
+
+Last, weirdly enough, we can remove our request handler for logout, as we
+included the `/logout` route as part of our security filter setup in
+`WebSecurity`. Navigating to `/logout` will invalidate the security session
+automatically.
+
+#### Update `EventCategoryController`
+
+To update `EventCategoryController`, we want to remove the reference to
+`authController` that is used to retrieve the current user. Instead we'll
+make a call to a `eventCategoryService` method that assumes the current
+user.
+
+In `EventCategoryController`, remove the field that references
+`AuthenticationController`, and update the `displayAllCategories` method to
+call `eventCategoryService.getAllCategoriesByCurrentUser()`, like below:
+
+```java
+    @GetMapping
+    public String displayAllCategories(Model model, HttpSession session) {
+        model.addAttribute("title", "All Categories");
+        model.addAttribute("categories", eventCategoryService.getAllCategoriesByCurrentUser());
+        return "eventCategories/index";
+    }
+```
+
+#### Update `EventController`
+
+Our goal for `EventController` will be the same as `EventCategoryController`
+--- we want to remove all references to `AuthenticationController` and the
+code that was necessary to grab the current user. Instead, we will make
+calls to `eventService` and `eventCategoryService` methods that assume the
+current user.
+
+First, **remove** the `AuthenticationController` autowired field.
+
+Next, we need one new method in `EventCategoryService` for retrieving a
+category by an id but also assuming the current user. We want to prevent
```

Comment:
```suggestion
Next, we need one new method in `EventCategoryService` for retrieving a
category that the current user created by the category id. We want to prevent
```